devito/finite_differences/differentiable.py

```diff
@@ -46,11 +47,48 @@ def time_order(self):
         return min([getattr(i, 'time_order', 100) or 100 for i in self._args_diff],
                    default=100)
 
+    @cached_property
+    def is_TimeDependent(self):
+        # Default False, True if anything is time dependant in the expression
+        return any(getattr(i, 'is_TimeDependent', False) for i in self._args_diff)
+
+    @cached_property
+    def is_VectorValued(self):
+        # Default False, True if is a vector valued expression
+        return any(getattr(i, 'is_VectorValued', False) for i in self._args_diff)
+
+    @cached_property
+    def is_TensorValued(self):
+        # Default False, True if is a tensor valued expression
+        return any(getattr(i, 'is_TensorValued', False) for i in self._args_diff)
+
+    @cached_property
+    def is_Function(self):
+        # Default False, True if is a Function
+        return any(getattr(i, 'is_Function', False) for i in self._args_diff)
+
+    @cached_property
+    def grid(self):
+        grids = {getattr(i, 'grid', None) for i in self._args_diff} - {None}
+        if len(grids) > 1:
+            warning("Expression contains multiple grids, returning first found")
+        return list(grids)[0]
```

Comment:
just noticed that this will fail if there's no grid attached. And currently, we still support grid-less Functions. So maybe it needs to be
```
try:
     return grids.pop()
except KeyError:
     raise ValueError(...)
```
?

Reply:
`grids.pop()` makes everything error, end up with weird stuff so keeping list.

I'm don't think can end up here with grid-less Function.

Reply:
what error?

> I'm don't think
is this enough ?

Reply:
>what error?
Unhashable type 'set'